<h1><center>Laboratorio 3: El Pandas no muerde 🐼</center></h1>

<center><strong>MDS7202: Laboratorio de Programación Científica para Ciencia de Datos</strong></center>

### Cuerpo Docente:

- Profesores: Ignacio Meza, Sebastián Tinoco
- Auxiliar: Catherine Benavides
- Ayudante: Nicolás Ojeda, Eduardo Moya

### Equipo: SUPER IMPORTANTE - notebooks sin nombre no serán revisados

- Nombre de alumno 1: Sergio Rehbein
- Nombre de alumno 2: Matías Cornejo

### **Link de repositorio de GitHub:**

#### **Link de repositorio de GitHub Matias:** https://github.com/s-kill/MDS7202

#### **Link de repositorio de GitHub Sergio:** https://github.com/sergiorehbein/MDS7201---Proyecto-de-Ciencia-de-Datos

### Reglas:

- **Grupos de 2 personas**
- Fecha de entrega: 7 días desde la publicación, 3 días de atraso con 1 punto de descuento c/u.
- Instrucciones del lab el viernes a las 16:15 en formato online. Asistencia no es obligatoria.
- <u>Prohibidas las copias</u>. Cualquier intento de copia será debidamente penalizado con el reglamento de la escuela.
- Tienen que subir el laboratorio a u-cursos Y a su repositorio de github. Labs que no estén en u-cursos no serán revisados. Recuerden que el repositorio también tiene nota.
- Cualquier duda fuera del horario de clases al foro. Mensajes al equipo docente serán respondidos por este medio.
- Pueden usar cualquer material del curso que estimen conveniente.

### Temas a tratar

- Introducción al manejo de datos tabulares por medio de la libreria `pandas`.

### Objetivos principales del laboratorio

- Entender, aplicar y aprovechar las ventajas que nos ofrece la libreria `pandas` para manejar datos tabulares.
- Aprender a utilizar la documentación de pandas (y en general, de cualquier librería de python) como también a buscar soluciones en la web.

> **Nota**: El laboratorio deberá ser desarrollado sin el uso indiscriminado de iteradores nativos de python (aka "for", "while"). La idea es que aprendan a exprimir al máximo las funciones optimizadas que nos entrega `pandas`, las cuales vale mencionar, son bastante más eficientes que los iteradores nativos sobre DataFrames.

In [1]:
##Gitclone

!git clone https://github.com/sergiorehbein/lab03-archivos.git


Cloning into 'lab03-archivos'...
remote: Enumerating objects: 6, done.
remote: Counting objects: 100% (6/6), done.
remote: Compressing objects: 100% (6/6), done.
remote: Total 6 (delta 1), reused 0 (delta 0), pack-reused 0
Receiving objects: 100% (6/6), 57.02 KiB | 4.07 MiB/s, done.
Resolving deltas: 100% (1/1), done.


### Librerias utiles para el Lab 😸

In [3]:
# Libreria Core del lab.
import numpy as np
import pandas as pd
from IPython.display import display

#Libreria para plotear
!pip install --upgrade plotly
import plotly.express as px

# Librerias utiles
from sklearn.preprocessing import StandardScaler

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 15.7/15.7 MB 17.5 MB/s eta 0:00:00
  Attempting uninstall: plotly
    Found existing installation: plotly 5.15.0
    Uninstalling plotly-5.15.0:
      Successfully uninstalled plotly-5.15.0


# 1. Rendimiento en Estudiantes 📚

Para este laboratorio deberán trabajar con el dataset `students_grades.json` el cual contiene información sobre el rendimiento de alumnos en distintos ramos.
El objetivo principal será el de realizar un trabajo de limpieza de prepración y exploración de datos utilizando exclusivamente `DataFrames` y sus métodos asociados.




In [ ]:
# Si usted está utilizando Colabolatory le puede ser útil este código para cargar los archivos.
try:
    from google.colab import drive
    drive.mount("/content/drive")
    path = 'Dirección donde tiene los archivos en el Drive'
except:
    print('Ignorando conexión drive-colab')

Ignorando conexión drive-colab


## 1. Carga e Inspección de Datos [1.0 Puntos]


<p align="center">
  <img src="https://drive.google.com/uc?export=view&id=1hZZaPQlZe4VupFOXt76_wjuUA5NpUhrw" width="300" height="300">
</p>

*Don Caguayo*, rector de la Universidad de la Cachaña, les comenta que actualmente la Universidad no tiene una subdirección de análisis de datos, pero que tiene un gran interés en formar una. Es por esto, que se contacta con ud. y su equipo para iniciar de forma exploratoria diversas tareas de análisis y consolidación de datos respectivos al rendimiento de sus alumnos.

Previo a su primera reunión con el rector, el rector le envía una copia de la base de datos, la cuál a usted y su equipo le encantaría examinar. Para esto, la primera tarea que debe realizar es abrir y cargar el dataset.


> **Notar que es sumamente importante el formato que posee cada columna** para su manipulación y que posiblemente ustedes deberán especificar algunos tipos de datos a mano. Ver la documentación del método `read_json` para mayor información.

A continuación se muestra un ejemplo de los dos primeros registros del dataset:

```json
[
  {
    "names": "Rita Courtney",
    "gender": "female",
    "race/ethnicity": "group B",
    "parental level of education": "some high school",
    "lunch": "standard",
    "test preparation course": "none",
    "math score": 37,
    "reading score": 46,
    "writing score": 46
  },
  {
    "names": "Madeline Fuller",
    "gender": "female",
    "race/ethnicity": "group C",
    "parental level of education": "some high school",
    "lunch": "standard",
    "test preparation course": "none",
    "math score": 124,
    "reading score": 142,
    "writing score": 142
  }
]
```





In [4]:
df_grades = pd.read_json('/content/lab03-archivos/students_grades.json')


# Convertir columnas numéricas al tipo adecuado
c_num = ['math score', 'reading score', 'writing score']
df_grades[c_num] = df_grades[c_num].apply(pd.to_numeric, errors='coerce')



<p align="center">
  <img src="https://i.giphy.com/media/4FQMuOKR6zQRO/giphy.webp">
</p>


Posteriormente, el rector les comenta que la sección de TI posee una base de datos con más de **875** filas, en donde cada fila representa un alumno matriculado. Sin embargo, cree fuertemente que existen errores en esta: solo tiene registro de **875** matriculas. Por este motivo, especula que se han realizado malas prácticas al registrar dichos datos.

En consecuencia, el rector les solicita a ustedes como tarea inicial que exploren la base de datos y encuentre posibles falencias dentro de sus registros. Para esto, usted debe estar al tanto que las notas están en una escala que van del **0 al 100** (porcentaje de logro) y **no es posible que un alumno haya dado más de una vez el mismo examen**. Cualquier alumno que posea un valor diferente a esta escala, debería ser eliminado de la base de datos.


> **Instrucciones**

**Cree una función** que tome un `DataFrame` y que le permita revisar si las diferentes variables que lo componen son coherentes y si es posible observar anomalías en estas.

**To-Do**
1. [ ] Imprimir un mensaje con las dimensiones del Dataframe.
2. [ ] Imprimir un mensaje con los nombres de las columnas que conforman el DataFrame.
3. [ ] Mostrar los primeros 5, los últimos 5 elementos del DataFrame y realice un muestreo de 5 elementos de forma aleatoria.
4. [ ] Mostrar una descripción rápida de las variables numéricas del DataFrame (utilice un método visto en clases).
5. [ ] Imprimir un mensaje con la cantidad de valores nulos (investigue el método `isna()` y `sum()` de los DataFrames).
6. [ ] Imprimir un mensaje que detalle la cantidad de valores unicos (investigue `nunique()`) .
7. [ ] Imprimir un mensaje con el total de filas que contengan nombres duplicados (investigue `duplicated()` y su parámetro `subset`).
8. [ ] Mostrar las filas que contengan nombres sean duplicados (investigue `duplicated()` y su parámetro `subset`).

Finalmente, responda las siguientes preguntas (comente esto en la celda ubicada debajo de **justificación**.):

9. [ ] ¿Qué anomalías es posible observar de los datos que componen el DataFrame? ¿Existen valores nulos o duplicados?


10. [ ] Su compañero ChatGPT le provee un código para que pueda entender los resultados de las pruebas desde un gráfico. ¿Qué puede interpretar de la distribución de las notas en el boxplot?

> **Nota:** Para imprmir utilice `print(...)`. Para mostrar `DataFrames`, utilice `display(...)` ya que este permite mostrar estos elementos como tablas.

> **Nota 2:** Utilice la documentación de [pandas](https://pandas.pydata.org/docs/index.html) como también StackOverflow para resolver cualquier problema o inconveniente.

> **Nota 3:** Si no ha trabajado con boxplot antes, el siguiente link puede serle de mucha ayuda: [boxplot](https://medium.com/dayem-siddiqui/understanding-and-interpreting-box-plots-d07aab9d1b6c)

**Respuesta:**

In [5]:
def exploratory_data_analysis(dataframe):
    # (1) Dimensiones
    filas, columnas = dataframe.shape
    print(f"1.- El DataFrame tiene {filas} filas y {columnas} columnas.")

    # (2) Columnas del DataFrame
    print(f"\n2.- El DataFrame está compuesto por las siguientes columnas: {dataframe.columns.tolist()}")

    # (3) Ejemplos de filas del DataFrame
    print("\n3.- Ejemplos de filas del DataFrame:")
    print("\nPrimeras 5 filas:")
    display(dataframe.head())
    print("\nÚltimas 5 filas:")
    display(dataframe.tail())
    print("\nMuestreo aleatorio de 5 filas:")
    display(dataframe.sample(5))

    # (4) Descripción numérica del DataFrame
    print("\n4.- Descripción numérica del DataFrame:")
    display(dataframe.describe(include='number'))

    # (5) Cantidad de valores nulos por columna
    print("\n5.- Cantidad de valores nulos por columna:")
    display(dataframe.isna().sum())

    # (6) Cantidad de valores únicos por columna
    print("\n6.- Cantidad de valores únicos por columna:")
    display(dataframe.nunique())

    # (7) Total de filas que contengan nombres duplicados
    total_duplicados = dataframe.duplicated(subset=['names']).sum()
    print(f"\n7.- Presenta un total de {total_duplicados} filas con nombres duplicados.")

    # (8) Las filas con nombres duplicados
    print("\n8.- Las filas con nombres duplicados son:")
    display(dataframe[dataframe.duplicated(subset=['names'], keep='first')])

    return

In [6]:
exploratory_data_analysis(df_grades)

1.- El DataFrame tiene 1400 filas y 9 columnas.

2.- El DataFrame está compuesto por las siguientes columnas: ['names', 'gender', 'race/ethnicity', 'parental level of education', 'lunch', 'test preparation course', 'math score', 'reading score', 'writing score']

3.- Ejemplos de filas del DataFrame:

Primeras 5 filas:


names  gender race/ethnicity parental level of education  \
0     Rita Courtney  female        group B            some high school   
1   Madeline Fuller  female        group C            some high school   
2     Daniel Rester    male        group E          associate's degree   
3  Charles Linstrom    male        group A           bachelor's degree   
4    Geraldine Hale  female        group A          associate's degree   

      lunch test preparation course  math score  reading score  writing score  
0  standard                    none        37.0           46.0           46.0  
1  standard                    none       124.0          142.0          142.0  
2  standard                    none       156.0          196.0          193.0  
3  standard               completed        80.0           78.0           81.0  
4  standard                    none       131.0          166.0          166.0


Últimas 5 filas:


names  gender race/ethnicity parental level of education  \
1395  Alexander Dillon    male        group D          associate's degree   
1396    Marina Zeigler  female        group C           bachelor's degree   
1397     Laurie Carter  female        group B            some high school   
1398      Joseph Mccoy    male        group D                some college   
1399      Amanda Perez  female        group A                 high school   

             lunch test preparation course  math score  reading score  \
1395      standard                    none       193.0          142.0   
1396  free/reduced               completed        66.0           74.0   
1397      standard               completed        54.0           61.0   
1398  free/reduced                    none       193.0          196.0   
1399      standard               completed        68.0           80.0   

      writing score  
1395          193.0  
1396           81.0  
1397           62.0  
1398          193.0  
1399           76.0


Muestreo aleatorio de 5 filas:


names  gender race/ethnicity parental level of education  \
1181     Jesus Currier    male        group B            some high school   
1152    Joyce Thornton  female        group B             master's degree   
897        Mark Spicer    male        group E          associate's degree   
411      Terry Wingerd  female        group E                 high school   
1022  Anthony Latouche    male        group B           bachelor's degree   

             lunch test preparation course  math score  reading score  \
1181      standard               completed        65.0           66.0   
1152  free/reduced               completed        58.0           76.0   
897       standard                    none        53.0           45.0   
411   free/reduced                    none        64.0           62.0   
1022  free/reduced                    none       193.0          196.0   

      writing score  
1181           62.0  
1152           78.0  
897            40.0  
411            68.0  
1022          193.0


4.- Descripción numérica del DataFrame:


math score  reading score  writing score
count  1360.000000    1360.000000    1353.000000
mean     89.538971      93.086029      91.852919
std      47.152422      46.385775      47.043570
min       0.000000      17.000000      10.000000
25%      59.000000      63.000000      62.000000
50%      72.000000      76.000000      75.000000
75%      97.250000     100.000000     100.000000
max     196.000000     196.000000     196.000000


5.- Cantidad de valores nulos por columna:


names                           0
gender                          0
race/ethnicity                  0
parental level of education     0
lunch                           0
test preparation course         0
math score                     40
reading score                  40
writing score                  47
dtype: int64


6.- Cantidad de valores únicos por columna:


names                          1153
gender                            2
race/ethnicity                    5
parental level of education       6
lunch                             2
test preparation course           2
math score                       96
reading score                    88
writing score                    92
dtype: int64


7.- Presenta un total de 247 filas con nombres duplicados.

8.- Las filas con nombres duplicados son:


names  gender race/ethnicity parental level of education  \
12         Rita Courtney  female        group B            some high school   
81    Elizabeth Quintero  female        group B                some college   
160    Michelle Gonzales  female        group B            some high school   
170        Carla Mendoza  female        group D          associate's degree   
176       Darlene Parker  female        group C          associate's degree   
...                  ...     ...            ...                         ...   
1385      James Crawford    male        group D             master's degree   
1390     Gloria Bradford  female        group B                 high school   
1393         Adam Romero    male        group C                 high school   
1395    Alexander Dillon    male        group D          associate's degree   
1398        Joseph Mccoy    male        group D                some college   

             lunch test preparation course  math score  reading score  \
12        standard                    none       193.0          193.0   
81        standard                    none        82.0           85.0   
160       standard               completed        60.0           70.0   
170   free/reduced                    none        43.0           60.0   
176       standard                    none        54.0           61.0   
...            ...                     ...         ...            ...   
1385      standard                    none       193.0          147.0   
1390      standard                    none       124.0           64.0   
1393      standard               completed       124.0          193.0   
1395      standard                    none       193.0          142.0   
1398  free/reduced                    none       193.0          196.0   

      writing score  
12            193.0  
81             87.0  
160            74.0  
170            58.0  
176            58.0  
...             ...  
1385          131.0  
1390          193.0  
1393           49.0  
1395          193.0  
1398          193.0  

[247 rows x 9 columns]

In [7]:
# Código del amigo ChatGPT (listo para usar)
fig = px.box(
    df_grades.melt(
        value_vars=["math score", "reading score", "writing score"],
        var_name="subject",
        value_name="score",
    ),
    y="score",
    color="subject",
)
fig.show()

**Justificación:**

```
Las notas están en una escala que van del **0 al 100**, por lo tanto se ven anomalías en los datos que deben ser limpiados.
```

**Respuesta Esperada:**

In [ ]:
exploratory_data_analysis(df_grades)

1.- El DataFrame tiene 1400 filas y 9 columnas

2.- El DataFrame esta compuesto por las siguientes columnas: ['names', 'gender', 'race/ethnicity', 'parental level of education', 'lunch', 'test preparation course', 'math score', 'reading score', 'writing score']

3.- Ejemplos de filas del DataFrame:

Primeras 5 filas:


names  gender race/ethnicity parental level of education  \
0     Rita Courtney  female        group B            some high school   
1   Madeline Fuller  female        group C            some high school   
2     Daniel Rester    male        group E          associate's degree   
3  Charles Linstrom    male        group A           bachelor's degree   
4    Geraldine Hale  female        group A          associate's degree   

      lunch test preparation course  math score  reading score  writing score  
0  standard                    none        37.0           46.0           46.0  
1  standard                    none       124.0          142.0          142.0  
2  standard                    none       156.0          196.0          193.0  
3  standard               completed        80.0           78.0           81.0  
4  standard                    none       131.0          166.0          166.0


Últimas 5 filas:


names  gender race/ethnicity parental level of education  \
1395  Alexander Dillon    male        group D          associate's degree   
1396    Marina Zeigler  female        group C           bachelor's degree   
1397     Laurie Carter  female        group B            some high school   
1398      Joseph Mccoy    male        group D                some college   
1399      Amanda Perez  female        group A                 high school   

             lunch test preparation course  math score  reading score  \
1395      standard                    none       193.0          142.0   
1396  free/reduced               completed        66.0           74.0   
1397      standard               completed        54.0           61.0   
1398  free/reduced                    none       193.0          196.0   
1399      standard               completed        68.0           80.0   

      writing score  
1395          193.0  
1396           81.0  
1397           62.0  
1398          193.0  
1399           76.0


Muestreo aleatorio de 5 filas:


names  gender race/ethnicity parental level of education  \
844    Jeffrey Korn    male        group C                 high school   
313      Johnny Lei    male        group A          associate's degree   
936     Martha Sova  female        group C            some high school   
1155    Bobby Lloyd    male        group C                some college   
453   Milton Chaput    male        group D                 high school   

             lunch test preparation course  math score  reading score  \
844       standard                    none         NaN          142.0   
313       standard                    none       124.0          124.0   
936   free/reduced                    none        43.0           53.0   
1155      standard                    none       193.0           49.0   
453       standard                    none        70.0           70.0   

      writing score  
844           196.0  
313             NaN  
936            53.0  
1155           42.0  
453            70.0


4.- Descripción numérica del Dataframe:


math score  reading score  writing score
count  1360.000000    1360.000000    1353.000000
mean     89.538971      93.086029      91.852919
std      47.152422      46.385775      47.043570
min       0.000000      17.000000      10.000000
25%      59.000000      63.000000      62.000000
50%      72.000000      76.000000      75.000000
75%      97.250000     100.000000     100.000000
max     196.000000     196.000000     196.000000


5.- Cantidad de valores nulos por columna:


names                           0
gender                          0
race/ethnicity                  0
parental level of education     0
lunch                           0
test preparation course         0
math score                     40
reading score                  40
writing score                  47
dtype: int64


6.- Cantidad de valores únicos por columna:


names                          1153
gender                            2
race/ethnicity                    5
parental level of education       6
lunch                             2
test preparation course           2
math score                       96
reading score                    88
writing score                    92
dtype: int64


7.- Presenta un total de 247 filas con nombres duplicados.

8.- Las filas con nombres duplicados son:


names  gender race/ethnicity parental level of education  \
12         Rita Courtney  female        group B            some high school   
81    Elizabeth Quintero  female        group B                some college   
160    Michelle Gonzales  female        group B            some high school   
170        Carla Mendoza  female        group D          associate's degree   
176       Darlene Parker  female        group C          associate's degree   
...                  ...     ...            ...                         ...   
1385      James Crawford    male        group D             master's degree   
1390     Gloria Bradford  female        group B                 high school   
1393         Adam Romero    male        group C                 high school   
1395    Alexander Dillon    male        group D          associate's degree   
1398        Joseph Mccoy    male        group D                some college   

             lunch test preparation course  math score  reading score  \
12        standard                    none       193.0          193.0   
81        standard                    none        82.0           85.0   
160       standard               completed        60.0           70.0   
170   free/reduced                    none        43.0           60.0   
176       standard                    none        54.0           61.0   
...            ...                     ...         ...            ...   
1385      standard                    none       193.0          147.0   
1390      standard                    none       124.0           64.0   
1393      standard               completed       124.0          193.0   
1395      standard                    none       193.0          142.0   
1398  free/reduced                    none       193.0          196.0   

      writing score  
12            193.0  
81             87.0  
160            74.0  
170            58.0  
176            58.0  
...             ...  
1385          131.0  
1390          193.0  
1393           49.0  
1395          193.0  
1398          193.0  

[247 rows x 9 columns]

## 2 Limpieza de Datos [1.0 puntos]




<p align="center">
  <img src="https://blogdatlas.files.wordpress.com/2020/07/datlas_mx_blog_limpieza_bds_data-scrubbing-service.gif">
</p>

En base a lo realizado anteriormente, realice una limpieza de los dataset. Asuma los supuestos señalados en la pregunta anterior para limpiar el `DataFrame` y cree una función que le permita limpiar los datos. Habiendo limpiado el `DataFrame`, compruebe nuevamente los datos con la función creada en la sección 1.1.

**To-Do:**

- [ ] Crear función que permita limpiar el dataset.
- [ ] Comprobar que no existen anomalías en los datos a traves de la función de la sección 1.1.

**Hint**

Puede ser útil los métodos ```.drop_duplicates()``` y ```.dropna()``` como también utilizar queries con máscaras booleanas.


**Respuesta Esperada:**

In [ ]:
df_grades = clean_data(df_grades)
df_grades

names  gender race/ethnicity parental level of education  \
0        Rita Courtney  female        group B            some high school   
3     Charles Linstrom    male        group A           bachelor's degree   
5          Brian Young    male        group C            some high school   
6       Howard Jimenez    male        group E            some high school   
7         Wayne Wilson    male        group B            some high school   
...                ...     ...            ...                         ...   
1392     Richard Young    male        group D                 high school   
1394     Wanda Russell  female        group B                 high school   
1396    Marina Zeigler  female        group C           bachelor's degree   
1397     Laurie Carter  female        group B            some high school   
1399      Amanda Perez  female        group A                 high school   

             lunch test preparation course  math score  reading score  \
0         standard                    none        37.0           46.0   
3         standard               completed        80.0           78.0   
5         standard                    none        73.0           66.0   
6         standard               completed        81.0           75.0   
7         standard               completed        94.0           86.0   
...            ...                     ...         ...            ...   
1392      standard                    none        69.0           75.0   
1394  free/reduced               completed        23.0           44.0   
1396  free/reduced               completed        66.0           74.0   
1397      standard               completed        54.0           61.0   
1399      standard               completed        68.0           80.0   

      writing score  
0              46.0  
3              81.0  
5              63.0  
6              76.0  
7              87.0  
...             ...  
1392           71.0  
1394           36.0  
1396           81.0  
1397           62.0  
1399           76.0  

[875 rows x 9 columns]

> Utilice la función creada en el punto 1.1 para verificar sus resultados.

In [8]:
def clean_data(df):
    # Eliminar filas duplicadas
    df = df.drop_duplicates(subset='names')

    # Eliminar filas con valores nulos
    df = df.dropna()


    # Filtrar puntajes que van de 0 a 100
    df = df[(df['math score'] >= 0) & (df['math score'] <= 100)]
    df = df[(df['reading score'] >= 0) & (df['reading score'] <= 100)]
    df = df[(df['writing score'] >= 0) & (df['writing score'] <= 100)]

    # Eliminar filas con mismos nombres pero puntajes distintos
    df = df.groupby('names').filter(lambda x: x[['math score', 'reading score', 'writing score']].nunique().max() == 1)

    return df



In [9]:
df_grades = clean_data(df_grades)
df_grades

names  gender race/ethnicity parental level of education  \
0        Rita Courtney  female        group B            some high school   
3     Charles Linstrom    male        group A           bachelor's degree   
5          Brian Young    male        group C            some high school   
6       Howard Jimenez    male        group E            some high school   
7         Wayne Wilson    male        group B            some high school   
...                ...     ...            ...                         ...   
1392     Richard Young    male        group D                 high school   
1394     Wanda Russell  female        group B                 high school   
1396    Marina Zeigler  female        group C           bachelor's degree   
1397     Laurie Carter  female        group B            some high school   
1399      Amanda Perez  female        group A                 high school   

             lunch test preparation course  math score  reading score  \
0         standard                    none        37.0           46.0   
3         standard               completed        80.0           78.0   
5         standard                    none        73.0           66.0   
6         standard               completed        81.0           75.0   
7         standard               completed        94.0           86.0   
...            ...                     ...         ...            ...   
1392      standard                    none        69.0           75.0   
1394  free/reduced               completed        23.0           44.0   
1396  free/reduced               completed        66.0           74.0   
1397      standard               completed        54.0           61.0   
1399      standard               completed        68.0           80.0   

      writing score  
0              46.0  
3              81.0  
5              63.0  
6              76.0  
7              87.0  
...             ...  
1392           71.0  
1394           36.0  
1396           81.0  
1397           62.0  
1399           76.0  

[875 rows x 9 columns]

**Justificación:**

```
1. Se elimina los duplicados por nombres, para que un estudiante no rinda más de dos veces una prueba.
2. Se eliminan los NA.
3. Se filtra por máximo y mínimo de notas (posible, 0 a 100).
```

## 3. Trabajando con Datos [2.0 puntos]

<p align="center">
  <img src="https://media3.giphy.com/media/YQitE4YNQNahy/giphy-downsized-large.gif">
</p>

### 3.1 Transformación de notas a "Nota Chilena" [0.3 Puntos]

En una de las reuniones con Don Caguayo este les comenta su intención de pasar todas sus notas a la "Escala Chilena". Para esto les facilita la siguiente ecuación:

$$Nota \, Chilena= \% \, de \, logro*0.06 + 1$$

Utilizando la ecuación entregada, cree una función que le permita pasar a la escala Chilena las notas del dataframe y entregue las notas redondeadas al segundo decimal.

**To-Do**

- [ ] Tranforme los valores a la escala chilena.

**Respuesta:**

In [10]:
def nota_chilena(dataframe_in, columna):

  for x in columna:
    dataframe_in[x] = dataframe_in[x].apply(lambda x: round(x  * 0.06 + 1, 2))
  return dataframe_in

In [11]:
df_grades=nota_chilena(df_grades, ['math score','reading score','writing score'])
df_grades

names  gender race/ethnicity parental level of education  \
0        Rita Courtney  female        group B            some high school   
3     Charles Linstrom    male        group A           bachelor's degree   
5          Brian Young    male        group C            some high school   
6       Howard Jimenez    male        group E            some high school   
7         Wayne Wilson    male        group B            some high school   
...                ...     ...            ...                         ...   
1392     Richard Young    male        group D                 high school   
1394     Wanda Russell  female        group B                 high school   
1396    Marina Zeigler  female        group C           bachelor's degree   
1397     Laurie Carter  female        group B            some high school   
1399      Amanda Perez  female        group A                 high school   

             lunch test preparation course  math score  reading score  \
0         standard                    none        3.22           3.76   
3         standard               completed        5.80           5.68   
5         standard                    none        5.38           4.96   
6         standard               completed        5.86           5.50   
7         standard               completed        6.64           6.16   
...            ...                     ...         ...            ...   
1392      standard                    none        5.14           5.50   
1394  free/reduced               completed        2.38           3.64   
1396  free/reduced               completed        4.96           5.44   
1397      standard               completed        4.24           4.66   
1399      standard               completed        5.08           5.80   

      writing score  
0              3.76  
3              5.86  
5              4.78  
6              5.56  
7              6.22  
...             ...  
1392           5.26  
1394           3.16  
1396           5.86  
1397           4.72  
1399           5.56  

[875 rows x 9 columns]

**Respuesta esperada:**

In [ ]:
df_grades

names  gender race/ethnicity parental level of education  \
0       Rita Courtney  female        group B            some high school   
1    Charles Linstrom    male        group A           bachelor's degree   
2         Brian Young    male        group C            some high school   
3      Howard Jimenez    male        group E            some high school   
4        Wayne Wilson    male        group B            some high school   
..                ...     ...            ...                         ...   
995     Richard Young    male        group D                 high school   
996     Wanda Russell  female        group B                 high school   
997    Marina Zeigler  female        group C           bachelor's degree   
998     Laurie Carter  female        group B            some high school   
999      Amanda Perez  female        group A                 high school   

            lunch test preparation course  math score  reading score  \
0        standard                    none        3.22           3.76   
1        standard               completed        5.80           5.68   
2        standard                    none        5.38           4.96   
3        standard               completed        5.86           5.50   
4        standard               completed        6.64           6.16   
..            ...                     ...         ...            ...   
995      standard                    none        5.14           5.50   
996  free/reduced               completed        2.38           3.64   
997  free/reduced               completed        4.96           5.44   
998      standard               completed        4.24           4.66   
999      standard               completed        5.08           5.80   

     writing score  
0             3.76  
1             5.86  
2             4.78  
3             5.56  
4             6.22  
..             ...  
995           5.26  
996           3.16  
997           5.86  
998           4.72  
999           5.56  

[1000 rows x 9 columns]

### 3.2 Obtención de promedio de alumnos [0.3 Puntos]

Con todas las notas adjuntadas, la prestigiosa Universidad De la Cachaña necesita obtener el promedio general de cada uno de los alumnos. Para esto se le solicita que calcule el promedio con cada una de las asignaturas y el promedio se anexe como una nueva columna al final del `DataFrame` original. Cabe señalar que las notas deben ser redondeadas al segundo decimal.

**To-Do:**
- [ ] Crear función que nos permite obtener el promedio de cada uno de los alumnos y anexar esta información en el `DataFrame` original.


**Respuesta:**

In [12]:
def alumnos_promedio(dataframe_in):
    # Calcular el promedio de cada alumno
    promedio_alumnos = dataframe_in[['math score', 'reading score', 'writing score']].mean(axis=1)

    # Redondear el promedio al segundo decimal
    promedio_alumnos = promedio_alumnos.round(2)

    # Anexar el promedio como una nueva columna al DataFrame original
    dataframe_in['GPA'] = promedio_alumnos

    return dataframe_in


In [14]:
df_grades = alumnos_promedio(df_grades)
df_grades.head()

names  gender race/ethnicity parental level of education  \
0     Rita Courtney  female        group B            some high school   
3  Charles Linstrom    male        group A           bachelor's degree   
5       Brian Young    male        group C            some high school   
6    Howard Jimenez    male        group E            some high school   
7      Wayne Wilson    male        group B            some high school   

      lunch test preparation course  math score  reading score  writing score  \
0  standard                    none        3.22           3.76           3.76   
3  standard               completed        5.80           5.68           5.86   
5  standard                    none        5.38           4.96           4.78   
6  standard               completed        5.86           5.50           5.56   
7  standard               completed        6.64           6.16           6.22   

    GPA  
0  3.58  
3  5.78  
5  5.04  
6  5.64  
7  6.34

**Respuesta Esperada:**

In [13]:
df_grades = alumnos_promedio(df_grades)
df_grades.head()

names  gender race/ethnicity parental level of education  \
0     Rita Courtney  female        group B            some high school   
3  Charles Linstrom    male        group A           bachelor's degree   
5       Brian Young    male        group C            some high school   
6    Howard Jimenez    male        group E            some high school   
7      Wayne Wilson    male        group B            some high school   

      lunch test preparation course  math score  reading score  writing score  \
0  standard                    none        3.22           3.76           3.76   
3  standard               completed        5.80           5.68           5.86   
5  standard                    none        5.38           4.96           4.78   
6  standard               completed        5.86           5.50           5.56   
7  standard               completed        6.64           6.16           6.22   

    GPA  
0  3.58  
3  5.78  
5  5.04  
6  5.64  
7  6.34

### 3.3 Reprobados y Destacados [0.9 Puntos, 0.3 cada DataFrame]

Contentos con su desempeño, la Universidad les solicita que obtengan los alumnos quienes han reprobado el año y los que han tenido desempeño de destacado. Para esto se considerará que un alumno ha reprobado el año cuando el promedio de notas es inferior a 4.0 en la Escala Chilena, y se considerará destacado si posee notas superiores a 5.8 en todas las asignaturas (o sea notas superiores a 5.8 en `math score`, `reading score` y `writing score`).

Cree una función que permita obtener tres Dataframes:
- El primero deberá poseer solamente los alumnos que han reprobado en la prestigiosa universidad.
- El segundo debe entregar un DataFrame similar al original, pero marcado con una "R" el promedio de notas de los alumnos reprobados.
- El tercero será un Data Frame con los alumnos destacados ordenados de mayor a menor.

**To-Do:**

- [ ] Obtener DataFrame con los alumnos que reprobaron el año.
- [ ] Obtener DataFrame similar al original, pero marcados con una "R" el atributo GPA de los alumnos reprobados.
- [ ] Obtener DataFrame con los alumnos destacados ordenados de mayor a menor.

**Respuesta:**

In [25]:
def reprobados_destacados(dataframe_in):
    df_reprobados = dataframe_in[dataframe_in['GPA']<4]

    df_R = dataframe_in.copy()
    df_R.loc[df_R['GPA']<4, 'GPA'] = 'R'

    df_destacados = dataframe_in.loc[(dataframe_in['math score']>5.8) & (dataframe_in['reading score']>5.8) & (dataframe_in['writing score']>5.8)]
    df_destacados = df_destacados.sort_values(by=['GPA'], ascending=False)

    return df_reprobados, df_R, df_destacados


In [26]:
dataframe1, dataframe2, dataframe3 = reprobados_destacados(df_grades)
dataframe1

names  gender race/ethnicity parental level of education  \
0       Rita Courtney  female        group B            some high school   
13        Mikki Amaya  female        group A                 high school   
26      Samuel Dougal    male        group C             master's degree   
55    Rodger Talavera    male        group B                 high school   
63     Morris Johnson    male        group A          associate's degree   
...               ...     ...            ...                         ...   
1371   Charles Harian    male        group C            some high school   
1373    Wendy Gunnell  female        group D            some high school   
1389      Diana Costa  female        group C                 high school   
1391      Amy Sieving  female        group C                 high school   
1394    Wanda Russell  female        group B                 high school   

             lunch test preparation course  math score  reading score  \
0         standard                    none        3.22           3.76   
13    free/reduced               completed        3.04           3.88   
26    free/reduced               completed        3.76           3.52   
55    free/reduced                    none        3.16           2.74   
63    free/reduced               completed        3.40           4.30   
...            ...                     ...         ...            ...   
1371  free/reduced               completed        3.70           4.12   
1373  free/reduced                    none        2.62           3.04   
1389  free/reduced                    none        3.46           3.76   
1391      standard                    none        2.74           2.74   
1394  free/reduced               completed        2.38           3.64   

      writing score   GPA  
0              3.76  3.58  
13             3.46  3.46  
26             3.76  3.68  
55             2.62  2.84  
63             4.18  3.96  
...             ...   ...  
1371           3.94  3.92  
1373           2.92  2.86  
1389           3.58  3.60  
1391           2.80  2.76  
1394           3.16  3.06  

[86 rows x 10 columns]

In [17]:
dataframe2

names  gender race/ethnicity parental level of education  \
0        Rita Courtney  female        group B            some high school   
3     Charles Linstrom    male        group A           bachelor's degree   
5          Brian Young    male        group C            some high school   
6       Howard Jimenez    male        group E            some high school   
7         Wayne Wilson    male        group B            some high school   
...                ...     ...            ...                         ...   
1392     Richard Young    male        group D                 high school   
1394     Wanda Russell  female        group B                 high school   
1396    Marina Zeigler  female        group C           bachelor's degree   
1397     Laurie Carter  female        group B            some high school   
1399      Amanda Perez  female        group A                 high school   

             lunch test preparation course  math score  reading score  \
0         standard                    none        3.22           3.76   
3         standard               completed        5.80           5.68   
5         standard                    none        5.38           4.96   
6         standard               completed        5.86           5.50   
7         standard               completed        6.64           6.16   
...            ...                     ...         ...            ...   
1392      standard                    none        5.14           5.50   
1394  free/reduced               completed        2.38           3.64   
1396  free/reduced               completed        4.96           5.44   
1397      standard               completed        4.24           4.66   
1399      standard               completed        5.08           5.80   

      writing score   GPA  
0              3.76     R  
3              5.86  5.78  
5              4.78  5.04  
6              5.56  5.64  
7              6.22  6.34  
...             ...   ...  
1392           5.26   5.3  
1394           3.16     R  
1396           5.86  5.42  
1397           4.72  4.54  
1399           5.56  5.48  

[875 rows x 10 columns]

In [18]:
dataframe3

names  gender race/ethnicity parental level of education  \
677       Richard Pryor    male        group E           bachelor's degree   
1352      Sandra Pompey  female        group E          associate's degree   
384   Elizabeth Beasley  female        group E           bachelor's degree   
475           Emma Gray  female        group E           bachelor's degree   
279        Emma Brasher  female        group D            some high school   
...                 ...     ...            ...                         ...   
949      Rolando Widmer    male        group C                 high school   
296   Lucille Stanphill  female        group C                 high school   
300        Steven Kelly    male        group D                some college   
1307        Peter Perez    male        group D             master's degree   
96          Harry Berry    male        group B          associate's degree   

         lunch test preparation course  math score  reading score  \
677   standard               completed        7.00           7.00   
1352  standard                    none        7.00           7.00   
384   standard                    none        7.00           7.00   
475   standard               completed        6.94           7.00   
279   standard               completed        6.82           7.00   
...        ...                     ...         ...            ...   
949   standard               completed        5.92           6.04   
296   standard                    none        5.86           6.04   
300   standard                    none        5.86           5.92   
1307  standard                    none        5.86           5.86   
96    standard               completed        5.86           5.92   

      writing score   GPA  
677            7.00  7.00  
1352           7.00  7.00  
384            7.00  7.00  
475            7.00  6.98  
279            7.00  6.94  
...             ...   ...  
949            5.92  5.96  
296            5.92  5.94  
300            6.04  5.94  
1307           6.04  5.92  
96             5.92  5.90  

[94 rows x 10 columns]

**Respuesta:**

In [19]:
dataframe1, dataframe2, dataframe3 = reprobados_destacados(df_grades)
dataframe1

names  gender race/ethnicity parental level of education  \
0       Rita Courtney  female        group B            some high school   
13        Mikki Amaya  female        group A                 high school   
26      Samuel Dougal    male        group C             master's degree   
55    Rodger Talavera    male        group B                 high school   
63     Morris Johnson    male        group A          associate's degree   
...               ...     ...            ...                         ...   
1371   Charles Harian    male        group C            some high school   
1373    Wendy Gunnell  female        group D            some high school   
1389      Diana Costa  female        group C                 high school   
1391      Amy Sieving  female        group C                 high school   
1394    Wanda Russell  female        group B                 high school   

             lunch test preparation course  math score  reading score  \
0         standard                    none        3.22           3.76   
13    free/reduced               completed        3.04           3.88   
26    free/reduced               completed        3.76           3.52   
55    free/reduced                    none        3.16           2.74   
63    free/reduced               completed        3.40           4.30   
...            ...                     ...         ...            ...   
1371  free/reduced               completed        3.70           4.12   
1373  free/reduced                    none        2.62           3.04   
1389  free/reduced                    none        3.46           3.76   
1391      standard                    none        2.74           2.74   
1394  free/reduced               completed        2.38           3.64   

      writing score   GPA  
0              3.76  3.58  
13             3.46  3.46  
26             3.76  3.68  
55             2.62  2.84  
63             4.18  3.96  
...             ...   ...  
1371           3.94  3.92  
1373           2.92  2.86  
1389           3.58  3.60  
1391           2.80  2.76  
1394           3.16  3.06  

[86 rows x 10 columns]

In [20]:
dataframe2

names  gender race/ethnicity parental level of education  \
0        Rita Courtney  female        group B            some high school   
3     Charles Linstrom    male        group A           bachelor's degree   
5          Brian Young    male        group C            some high school   
6       Howard Jimenez    male        group E            some high school   
7         Wayne Wilson    male        group B            some high school   
...                ...     ...            ...                         ...   
1392     Richard Young    male        group D                 high school   
1394     Wanda Russell  female        group B                 high school   
1396    Marina Zeigler  female        group C           bachelor's degree   
1397     Laurie Carter  female        group B            some high school   
1399      Amanda Perez  female        group A                 high school   

             lunch test preparation course  math score  reading score  \
0         standard                    none        3.22           3.76   
3         standard               completed        5.80           5.68   
5         standard                    none        5.38           4.96   
6         standard               completed        5.86           5.50   
7         standard               completed        6.64           6.16   
...            ...                     ...         ...            ...   
1392      standard                    none        5.14           5.50   
1394  free/reduced               completed        2.38           3.64   
1396  free/reduced               completed        4.96           5.44   
1397      standard               completed        4.24           4.66   
1399      standard               completed        5.08           5.80   

      writing score   GPA  
0              3.76     R  
3              5.86  5.78  
5              4.78  5.04  
6              5.56  5.64  
7              6.22  6.34  
...             ...   ...  
1392           5.26   5.3  
1394           3.16     R  
1396           5.86  5.42  
1397           4.72  4.54  
1399           5.56  5.48  

[875 rows x 10 columns]

In [21]:
dataframe3

names  gender race/ethnicity parental level of education  \
677       Richard Pryor    male        group E           bachelor's degree   
1352      Sandra Pompey  female        group E          associate's degree   
384   Elizabeth Beasley  female        group E           bachelor's degree   
475           Emma Gray  female        group E           bachelor's degree   
279        Emma Brasher  female        group D            some high school   
...                 ...     ...            ...                         ...   
949      Rolando Widmer    male        group C                 high school   
296   Lucille Stanphill  female        group C                 high school   
300        Steven Kelly    male        group D                some college   
1307        Peter Perez    male        group D             master's degree   
96          Harry Berry    male        group B          associate's degree   

         lunch test preparation course  math score  reading score  \
677   standard               completed        7.00           7.00   
1352  standard                    none        7.00           7.00   
384   standard                    none        7.00           7.00   
475   standard               completed        6.94           7.00   
279   standard               completed        6.82           7.00   
...        ...                     ...         ...            ...   
949   standard               completed        5.92           6.04   
296   standard                    none        5.86           6.04   
300   standard                    none        5.86           5.92   
1307  standard                    none        5.86           5.86   
96    standard               completed        5.86           5.92   

      writing score   GPA  
677            7.00  7.00  
1352           7.00  7.00  
384            7.00  7.00  
475            7.00  6.98  
279            7.00  6.94  
...             ...   ...  
949            5.92  5.96  
296            5.92  5.94  
300            6.04  5.94  
1307           6.04  5.92  
96             5.92  5.90  

[94 rows x 10 columns]

### 3.4 Quintiles de Rendimiento [0.5 Puntos]

El rector, intrigado por saber si existe una relación entre el nivel educacional de los padres y las notas de sus hijos les plantea estudiar dicha relación. Para esto, les solicita que desde el `DataFrame` que posee a todos los alumnos (`DataFrame` obtenido tras la limpieza), generen una nueva columna que clasifique en quintiles de rendimiento para sus alumnos: `['Bajísimo', 'Bajo', 'Medio', 'Alto', 'Muy alto']`. (investigar `qcut()`)

Hecho esto, grafique a través de un scatter plot el nivel educacional de los padres v/s el promedio de los alumnos, utilice los cuantiles generados anteriormente para mostrar en que cuantil se encuentra cada uno de los datos.

Realizado el grafico, responda la siguiente pregunta: **¿Es posible visualizar alguna tendencia en el gráfico?.**



**Respuesta:**

In [27]:
# Código para quintiles
df_clean = df_grades.copy()

"""
NOTA: La variable del dataframe que contenga los quintiles debe llamarse
'GPA cuantiles' para que el gráfico funcione.
"""

df_clean['GPA cuantiles'] = pd.qcut(df_clean['GPA'], q=5, labels=['Bajísimo', 'Bajo', 'Medio', 'Alto', 'Muy alto'])

In [29]:
# Código de gráfico del amigo ChatGPT
fig = px.scatter(
    df_clean.sort_values(by="GPA"),
    x="GPA",
    y="parental level of education",
    hover_data=["gender"],
    color="GPA cuantiles",
    title="GPA por Nivel de educación de los padres",
    color_discrete_sequence=["tomato", "orange", "yellow", "greenyellow", "green"]
)

fig

## 4. Operaciones con tablas [2.0 puntos]

<p align="center">
  <img src="https://media.giphy.com/media/v1.Y2lkPTc5MGI3NjExYWx5dTBkMGhhOTJmeGQzbDV5M21rNTN1NWVwMm1qbHd3cjh3Nmk4NSZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/13HBDT4QSTpveU/giphy.gif">
</p>



### 4.1 Carga de datos y concatenar dataset [0.2 Puntos]

Ya finalizado en análisis inicial, ud. y su equipo le entregaron a *Don Caguayo* (rector de la Universidad de la Cachaña) tanto los resultados del análisis como también la base de datos limpia y lista para ser almacenada. Dada la ingente cantidad de los datos, el equipo de TI de la universidad resolvió separar el dataset en dos bases de datos distintas (lo que según argumentan ellos, permitiría hacer agregaciones de forma más eficiente).

Gracias a la excelente labor de ud. y su equipo en el análisis previo, el rector le solicita continuar el trabajo con una nueva batería de análisis. Por este motivo, la sección de TI les entrega nuevamente los datos. Sin embargo, argumentan que dada una escazes de personal, solo le entregarán *dumps* (copias) de cada base de datos y su equipo deberá unir las bases de datos. Los datos se encuentran en los siguiente archivos `.json`: `students_grades_1.json` y `students_grades_2.json`.


Por ende, ud. y su equipo deciden que la primera tarea se centrará en cargar estos datos y unirlos.

**No se preocupe por la limpieza ni transformar el tipo de datos de las columnas, ni tampoco transformar a notas chilenas**, recuerde que anteriormente ya se encargo de este tema.


Hint: investigue el método `concat`

In [31]:
df_sg1 = pd.read_json("/content/lab03-archivos/students_grades_1.json")
df_sg2 = pd.read_json("/content/lab03-archivos/students_grades_2.json")

df_sg = pd.concat([df_sg1,df_sg2])
df_sg

names  gender race/ethnicity parental level of education  \
0       Rita Courtney  female        group B            some high school   
1    Charles Linstrom    male        group A           bachelor's degree   
2         Brian Young    male        group C            some high school   
3      Howard Jimenez    male        group E            some high school   
4        Wayne Wilson    male        group B            some high school   
..                ...     ...            ...                         ...   
470     Richard Young    male        group D                 high school   
471     Wanda Russell  female        group B                 high school   
472    Marina Zeigler  female        group C           bachelor's degree   
473     Laurie Carter  female        group B            some high school   
474      Amanda Perez  female        group A                 high school   

            lunch test preparation course  math score  reading score  \
0        standard                    none        3.22           3.76   
1        standard               completed        5.80           5.68   
2        standard                    none        5.38           4.96   
3        standard               completed        5.86           5.50   
4        standard               completed        6.64           6.16   
..            ...                     ...         ...            ...   
470      standard                    none        5.14           5.50   
471  free/reduced               completed        2.38           3.64   
472  free/reduced               completed        4.96           5.44   
473      standard               completed        4.24           4.66   
474      standard               completed        5.08           5.80   

     writing score  
0             3.76  
1             5.86  
2             4.78  
3             5.56  
4             6.22  
..             ...  
470           5.26  
471           3.16  
472           5.86  
473           4.72  
474           5.56  

[875 rows x 9 columns]

### 4.2 Análisis por Nivel Educacional Etnia de los Padres [0.8 Puntos]

El rector, basado en su experiencia, cree fuertemente que el nivel educacional y la etnia de los padres influyen en las notas que obtienen sus hijos.
Como científicos de datos, ud. y su equipo creen que deben encontrar evidencia para confirmar o refutar la hipótesis del rector.

Para esto, deciden generar dos análisis: una tabla de resumen por una parte y gráficos de caja por otro.


**Tabla de Resumen [0.5 Puntos] **


Para generar la tabla de resumen:

- [ ] Calcular el promedio de las notas y guardarlo en una variable `GPA` (grade point average).
- [ ] Hacer una simplificación a través de un mapeo (investigar el método `map()`) de la variable `parental level of education`  según la siguiente conversión:

        some high school -> school
        some college -> school
        high school -> school
        bachelor's degree -> college
        associate's degree -> college
        master's degree -> postgraduate

  Los resultados de este mapeo deben ser guardados en la columna `simple parental level of education`.


- [ ] Agregar según 2 niveles: `race/ethnicity` y `simple parental level of education` para obtener el promedio de las notas.
- [ ] Agregar según 2 niveles: `race/ethnicity` y `simple parental level of education` para obtener un conteo de los alumnos en cada grupo y agregarlos como una nueva fila count.
- [ ] Obtener el porcentaje de alumnos con respecto al total. Los porcentajes deben ser strings que contienen la frecuencia de cada grupo con respecto al total y deben ser terminados en '%'.

Utilizar la tabla de resultados esperados como guía para desarrollar este punto.


**Análisis [0.3 Puntos]**

Finalmente, responda las siguientes preguntas sobre el gráfico de caja que viene implementado a continuación. En este se muestra el GPA con respecto al nivel educacional, donde la variable de color es la etnicidad.

- [ ]  ¿Hay alguna diferencia entre los grupos graficados tanto para el nivel educacional de los padres como también para la etnicidad?


In [32]:
# Calcular el promedio de las notas y guardarlo en una variable GPA
df_sg = alumnos_promedio(df_sg)

#Hacer una simplificación a través de un mapeo de la variable
parental_level_dic = {
    "some high school" : "school",
    "some college" : "school",
    "high school" : "school",
    "bachelor's degree" : "college",
    "associate's degree" : "college",
    "master's degree" : "postgraduate"
}

df_sg['simple parental level of education'] = df_sg['parental level of education'].map(parental_level_dic)

#Agregar según 2 niveles: race/ethnicity y simple parental level of education para obtener el promedio de las notas.
tabla_resumen = df_sg.groupby(['race/ethnicity', 'simple parental level of education'])[['math score','reading score', 'writing score', 'GPA']].mean().reset_index()

# Agregar según 2 niveles: race/ethnicity y simple parental level of education para obtener un conteo de los alumnos en cada grupo y agregarlos como una nueva fila count.
df_count = df_sg.groupby(['race/ethnicity', 'simple parental level of education']).size().reset_index(name='count')
tabla_resumen['count'] = df_count['count']

#Obtener el porcentaje de alumnos con respecto al total. Los porcentajes deben ser strings que contienen la frecuencia de cada grupo con respecto al total y deben ser terminados en '%'.
N = len(df_sg)
tabla_resumen['porcentage'] = (100 * tabla_resumen['count'] / N).round(2).astype(str) + '%'

In [33]:
fig = px.box(
    tabla_resumen,
    x="simple parental level of education",
    y="GPA",
    color="race/ethnicity",
    hover_data=["GPA"],
)
fig.show()

**Respuesta**:
Observando el gráfico, se observan claramentes los clusters que se forman a nivel de educacion, donde a priori se observa que los promedios de 'postgraduate' son generalmente mayores que 'college', y este último, generalmente mayores que 'school'. Por otro lado, se observa la dominancia del 'group E' para los tres niveles de educación. Se confirma la hipotesis del rector.

**Resultado Esperado de la tabla**

|    | race/ethnicity   | simple parental level of education   |   math score |   reading score |   writing score |   GPA |   count | percentage   |
|---:|:-----------------|:-------------------------------------|-------------:|----------------:|----------------:|------:|--------:|:-------------|
|  0 | group A          | college                              |         4.74 |            5    |            4.89 |  4.88 |      24 | 2.74 %       |
|  1 |           | postgraduate                         |         4.69 |            5.23 |            5.35 |  5.09 |       2 | 0.23 %       |
|  2 |           | school                               |         4.57 |            4.73 |            4.56 |  4.62 |      51 | 5.83 %       |
|  3 | group B          | college                              |         5.07 |            5.26 |            5.19 |  5.18 |      54 | 6.17 %       |
|  4 |           | postgraduate                         |         4.91 |            5.69 |            5.55 |  5.38 |       5 | 0.57 %       |
|  5 |           | school                               |         4.69 |            4.89 |            4.76 |  4.78 |     107 | 12.23 %      |
|  6 | group C          | college                              |         5.02 |            5.37 |            5.35 |  5.25 |     102 | 11.66 %      |
|  7 |           | postgraduate                         |         4.92 |            5.14 |            5.1  |  5.06 |      15 | 1.71 %       |
|  8 |           | school                               |         4.76 |            5.02 |            4.92 |  4.9  |     155 | 17.71 %      |
|  9 | group D          | college                              |         5.11 |            5.25 |            5.25 |  5.2  |      70 | 8.0 %        |
| 10 |           | postgraduate                         |         5.22 |            5.54 |            5.73 |  5.5  |      20 | 2.29 %       |
| 11 |           | school                               |         5.02 |            5.13 |            5.11 |  5.09 |     149 | 17.03 %      |
| 12 | group E          | college                              |         5.54 |            5.45 |            5.45 |  5.48 |      52 | 5.94 %       |
| 13 |          | postgraduate                         |         5.54 |            6.03 |            5.89 |  5.82 |       6 | 0.69 %       |
| 14 |          | school                               |         5.4  |            5.31 |            5.16 |  5.29 |      63 | 7.2 %        |

### 4.3 Joins, joins y más joins [1 punto]

Mientras le notificaba por videollamada los resultados de sus descubrimientos a Don Caguayo, un exaltado practicante del area de TI entra a la reunión y les informa que ha encontrado una nueva base de datos que cuenta con las notas de dos asignaturas (en escala chilena): historia y ciencias.
Para más remate, antes de huir, el practicante les cuenta que este dataframe lamentablemente contiene nuevamente los alumnos de los registros corruptos que ud. y su equipo filtraron en el análisis anterior.


El rector (evidentemente molesto por la situación) les ruega incluir estos datos (vaciados en el archivo other_grades.csv) al estudio original(`students_grades.csv`).


Para esto, carge el archivo `other_grades.csv` y busque la forma de unir ambos DataFrames, de tal manera que las columnas de `history score` y `science score` se anexen al final del DataFrame original. **NO LIMPIE LOS DATOS**, si no que explore los distintos tipos de merge para encontrar el mas situable para su situación (y así evitar buscar duplicados).

**To-Do**


- [ ] Cargar el `other_grades.csv`
- [ ] Unir `df_grades` con `other_grades.csv` usando outer join y explique el resultado.
- [ ] Unir `df_grades` con `other_grades.csv` usando left join y explique el resultado.
- [ ] Unir `df_grades` con `other_grades.csv` usando right join y explique el resultado.
- [ ] Unir `df_grades` con `other_grades.csv` usando inner join y explique el resultado.
- [ ] Defina cuál join es el que utilizará para generar el nuevo DataFrame.

> **Hint**: Puede explicar los resultados del merge a través de la cantidad de filas resultantes y los valores que estas contienen.

In [35]:
#  Cargar el other_grades.csv
df_og = pd.read_csv("/content/lab03-archivos/other_grades.csv")

#Unir df_grades con other_grades.csv usando outer join y explique el resultado.
df_outer_join = pd.merge(df_grades, df_og, how='outer', on='names')

#Unir df_grades con other_grades.csv usando left join y explique el resultado.
df_left_join = pd.merge(df_grades, df_og, how='left', on='names')

#Unir df_grades con other_grades.csv usando right join y explique el resultado.
df_right_join = pd.merge(df_grades, df_og, how='right', on='names')

#Unir df_grades con other_grades.csv usando inner join y explique el resultado.
df_inner_join = pd.merge(df_grades, df_og, how='inner', on='names')

In [36]:
# Imprimir la cantidad de filas y columnas de cada DataFrame
print("df_grades:")
print("Filas:", df_grades.shape[0])
print("Columnas:", df_grades.shape[1])
print("\n")

print("df_og:")
print("Filas:", df_og.shape[0])
print("Columnas:", df_og.shape[1])
print("\n")

print("df_outer_join:")
print("Filas:", df_outer_join.shape[0])
print("Columnas:", df_outer_join.shape[1])
print("\n")

print("df_left_join:")
print("Filas:", df_left_join.shape[0])
print("Columnas:", df_left_join.shape[1])
print("\n")

print("df_right_join:")
print("Filas:", df_right_join.shape[0])
print("Columnas:", df_right_join.shape[1])
print("\n")

print("df_inner_join:")
print("Filas:", df_inner_join.shape[0])
print("Columnas:", df_inner_join.shape[1])
print("\n")

df_grades:
Filas: 875
Columnas: 10


df_og:
Filas: 1000
Columnas: 3


df_outer_join:
Filas: 1000
Columnas: 12


df_left_join:
Filas: 875
Columnas: 12


df_right_join:
Filas: 1000
Columnas: 12


df_inner_join:
Filas: 875
Columnas: 12




Se puede checkear que filas y columnas coinciden según el tipo de ejercicio.

# Conclusión


Eso ha sido todo para el lab de hoy, recuerden que el laboratorio tiene un plazo de entrega de una semana y que **los días de atraso no se pueden utilizar para entregas de lab** solo para tareas. Cualquier duda del laboratorio, no duden en contactarnos por mail o U-cursos.

<p align="center">
  <img src="https://media1.tenor.com/images/2f7df460196625edc45b8380d158e5ee/tenor.gif?itemid=16669403">
</p>